In [12]:
#导入数据，让我们使用digit recognizor数据来一展身手
import numpy as np
import pandas as pd

data = pd.read_csv(r"C:\Users\ZGY\Desktop\机器学习\train.csv\train.csv")#有情况会出现，内存不足以显示这么大矩阵的情况，这时用ubuntu进行调整。
X = data.iloc[:,1:]#抛去第0列的索引。
y = data.iloc[:,0]#这是索引

 
"""
这个数据量相对夸张，如果使用支持向量机和神经网络，很可能会直接跑不出来。使用KNN跑一次大概需要半个小时。
用这个数据举例，能更够体现特征工程的重要性。
"""
 
X.shape

MemoryError: Unable to allocate array with shape (785, 42000) and data type int64

In [4]:
from sklearn.feature_selection import VarianceThreshold
selector = VarianceThreshold()#实例化，不填参数默认删除方差为0

In [5]:
X_var0 = selector.fit_transform(X)#获取删除不合格特征之后的新特征矩阵
 
#也可以直接写成 X = VairanceThreshold().fit_transform(X)
 
X_var0.shape

(42000, 708)

### 可以看见，我们已经删除了方差为0的特征，但是依然剩下了708多个特征，明显还需要进一步的特征选择。然而，如果我们知道我们需要多少个特征，方差也可以帮助我们将特征选择一步到位。比如说，我们希望留下一半的特征，那可以设定一个让特征总数减半的方差阈值，只要找到特征方差的中位数，再将这个中位数作为参数threshold的值输入就好了：


In [6]:
import numpy as np
X_fsvar = VarianceThreshold(np.median(X.var().values)).fit_transform(X)
#X.var()：读取每一列的方差值。
#X.var().values:提取所有列的方差。
X.var().values

array([0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       3.22760590e-01, 2.64689018e+00, 1.11085714e+00, 1.92857143e-03,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       6.09523810e-03, 5.41186213e-02, 1.28065733e+00, 5.33793150e+00,
       9.74592647e+00, 1.06219147e+01, 2.49289953e+01, 3.37671294e+01,
       3.59425660e+01, 3.49309575e+01, 3.62943342e+01, 3.27911079e+01,
       3.04237672e+01, 2.95547817e+01, 2.00568398e+01, 1.19545659e+01,
       8.60419724e+00, 2.55465685e+00, 1.60530842e+00, 9.29377602e-01,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
      

In [7]:
 
np.median(X.var().values)
 
X_fsvar.shape

(42000, 392)

### 当特征是二分类时，特征的取值就是伯努利随机变量，这些变量的方差可以计算为：其中X是特征矩阵，p是二分类特征中的一类在这个特征中所占的概率。


In [8]:
#若特征是伯努利随机变量，假设p=0.8，即二分类特征中某种分类占到80%以上的时候删除特征
X_bvar = VarianceThreshold(.8 * (1 - .8)).fit_transform(X)
X_bvar.shape


MemoryError: Unable to allocate array with shape (42000, 784) and data type float64

### 随机森林方差过滤前后的变化

In [ ]:
from sklearn.ensemble import RandomForestClassifier as RFC
from sklearn.neighbors import KNeighborsClassifier as KNN
from sklearn.model_selection import cross_val_score
import numpy as np

In [10]:
%%timeit
cross_val_score(RFC(n_estimators=10,random_state=0),X,y,cv=5).mean()

MemoryError: Unable to allocate array with shape (784, 33600) and data type int64